docs/decisions/0017-reimplement-asset-processing.rst

```diff
@@ -0,0 +1,268 @@
+   :header-rows: 1
+
+   * - Pain Point
+     - Potential solution(s)
+
+   * - edx-platform Docker images are too large and/or take too long to build.
+     - Switch from large, legacy tooling packages (such as libsass-python and paver) to industry standard, pre-compiled ones (like node-sass or dart-sass). Remove unnecessary & slow calls to Django management commands.
+
+   * - edx-platform Docker image layers seem to be rebuilt more often than they should.
+     - Remove all Python dependencies from the static asset build process, such that changes to Python code or requirements do not always have to result in a static asset rebuild.
+
+   * - In Tutor, using a local copy of edx-platform overwrites the Docker image's pre-installed node_modules and pre-built static assets, requiring developers to reinstall & rebuild in order to get a working platform.
+     - Better parameterize the input and output paths edx-platform asset build, such that it may `search for node_modules outside of edx-platform <https://github.com/openedx/wg-developer-experience/issues/150> and `generate assets outside of edx-platform <https://github.com/openedx/wg-developer-experience/issues/151>`.
+
+All of these potential solutions would involve refactoring or entirely replacing parts of the current asset processing system.
+
+
+Decision
+********
+
+We will largely reimplement edx-platform's asset processing system. We will aim to:
+
+* Use well-known, npm-installed frontend tooling wherever possible.
+* When bespoke processing is required, use standard POSIX tools like Bash.
+* When Django/Python is absolutely required, contain its impact so that the rest of the system remains Python-free.
+* Avoid unnecessary indirection or abstraction. For this task, extensibility is a non-goal, and simplicity is a virtue.
+* Provide a clear migration path from the old system to the new one.
+* Enable the future removal of as much legacy frontend tooling code as possible.
+
+Consequences
+************
+
+Reimplementation Specification
+==============================
+
+The three top-level edx-platform asset processing actions are *build*, *collect*, and *watch*. The build action can be further broken down into five stages. Here is how those actions and stages will be reimplemented:
+
+
+.. list-table::
+   :header-rows: 1
+
+   * - Description
+     - Old implementation
+     - New implementation
+
+   * - **Build: All stages.** Compile, generate, copy, and otherwise process static assets so that they can be used by the Django webserver or collected elsewhere. For many Web applications, all static asset building would be coordinated via Webpack or another NPM-managed tool. Due to the age of edx-platform and its legacy XModule and Comprehensive Theming systems, though, there are five stages which need to be performed in a particular order.
+
+     - ``paver update_assets --skip-collect``
+
+       A Python-defined task that calls out to each build stage.
+
+     - ``assets/build.sh``
```

Comment:
I'd originally come up with `assets/build.sh` as the script name because I thought it'd be good to have a new top-level `assets/` folder in edx-platform to hold:
1. build.sh
2. a "collect.sh" script
3. Python requirements related to asset building
4. anything else asset-related

Now that I've fleshed everything out, though, I realized that we don't need a collect.sh script (the collectstatic management commands is sufficient) and we won't need any special Python requirements for the asset build (it'll all be Bash + npm dependencies).

Given that, I think I should get rid of the `assets/` directory, and rename the script `scripts/build-assets.sh`. Any objections?

Reply:
No complaints from me either.

Reply:
This is done.